semtypes/src/main/java/io/ballerina/semtype/definition/MappingDefinition.java

```diff
@@ -31,12 +44,88 @@
  */
 public class MappingDefinition implements Definition {
 
+    private RecAtom roRec = null;
+    private RecAtom rwRec = null;
+    private SemType semType = null;
+
     @Override
     public SemType getSemType(Env env) {
-        throw new AssertionError();
+        SemType s = this.semType;
+        if (s == null) {
+            RecAtom ro = env.recMappingAtom();
+            RecAtom rw = env.recMappingAtom();
+            this.roRec = ro;
+            this.rwRec = rw;
+            return createSemType(env, ro, rw);
+        } else {
+            return s;
+        }
+    }
+
+    public SemType define(Env env, List<Field> fields, SemType rest) {
+        SplitField sfh = splitFields(fields);
+        String[] nameArray = new String[sfh.names.size()];
+        SemType[] typeArray = new SemType[sfh.types.size()];
+        MappingAtomicType rwType = MappingAtomicType.from(sfh.names.toArray(nameArray),
```

Comment:
Shall we do the usual `new String[]{}`, it's usually recognized by JVM JIT https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_introduction